mithril-common/src/resource_pool.rs

```diff
@@ -0,0 +1,320 @@
+pub struct ResourcePoolItem<'a, T: Send + Sync> {
+    resource_pool: &'a ResourcePool<T>,
+    discriminant: u64,
+    resource: Option<T>,
+}
+
+impl<'a, T: Send + Sync> ResourcePoolItem<'a, T> {
+    /// Create a new resource pool item
+    pub fn new(resource_pool: &'a ResourcePool<T>, resource: T) -> Self {
+        let discriminant = *resource_pool.discriminant.lock().unwrap();
+        Self {
+            resource_pool,
+            discriminant,
+            resource: Some(resource),
+        }
+    }
+
+    /// Get the discriminant of the resource pool item
+    pub fn discriminant(&self) -> u64 {
+        self.discriminant
+    }
+
+    /// Get a reference to the inner resource
+    pub fn resource(&self) -> &T {
+        self.resource.as_ref().unwrap()
+    }
+
+    /// Take the inner resource
+    pub fn into_inner(&mut self) -> T {
+        self.resource.take().unwrap()
+    }
+}
+
+impl<T: Send + Sync> Deref for ResourcePoolItem<'_, T> {
+    type Target = T;
+
+    fn deref(&self) -> &T {
+        self.resource.as_ref().unwrap()
+    }
+}
+
+impl<T: Send + Sync> DerefMut for ResourcePoolItem<'_, T> {
+    fn deref_mut(&mut self) -> &mut T {
+        self.resource.as_mut().unwrap()
+    }
+}
+
+impl<T: Send + Sync> Drop for ResourcePoolItem<'_, T> {
+    fn drop(&mut self) {
+        if self.resource.is_some() {
+            let resource = self.into_inner();
+            let _ = self
+                .resource_pool
+                .give_back_resource(resource, self.discriminant);
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::time::Duration;
+
+    use super::*;
+
+    #[test]
+    fn test_resource_pool_acquire_returns_resource_when_available() {
+        let pool_size = 10;
+        let resources_expected: Vec<String> = (0..pool_size).map(|i| i.to_string()).collect();
+        let pool = ResourcePool::<String>::new(pool_size, resources_expected.clone());
+
+        let mut resources_items = vec![];
+        for _ in 0..pool_size {
+            let resource_item = pool.acquire_resource(Duration::from_millis(10)).unwrap();
+            resources_items.push(resource_item);
+        }
+        let resources_result = resources_items
+            .iter_mut()
+            .map(|resource_item| resource_item.resource().to_owned())
+            .collect::<Vec<_>>();
+
+        assert_eq!(resources_expected, resources_result);
+        assert_eq!(pool.count().unwrap(), 0);
+    }
+
+    #[tokio::test]
+    async fn test_resource_pool_acquire_locks_until_timeout_when_no_resource_available() {
+        let pool_size = 10;
+        let resources_expected: Vec<String> = (0..pool_size).map(|i| i.to_string()).collect();
+        let pool = ResourcePool::<String>::new(pool_size, resources_expected.clone());
+
+        let mut resources_items = vec![];
+        for _ in 0..pool_size {
+            let resource_item = pool.acquire_resource(Duration::from_millis(10)).unwrap();
+            resources_items.push(resource_item);
```

Comment:
The push seems useless because resources_items is never use while it's needed otherwise the resource will go back to the pool.
Can this be more explicit ?

Reply:
I guess that's basic feature of the language, but I'll make it explicit with a comment.